File: FactoryMaze/src/com/ash/maze/Room.java
package com.ash.maze;

import java.util.HashMap;

public class Room implements MapSite {
    protected int roomNumber;
    protected HashMap<Direction, MapSite> sides;

    public Room() {
        this(0);
    }

    public Room(int roomNumber) {
        this.roomNumber = roomNumber;
        this.sides = new HashMap<>(4);
    }

    public int number() {
        return this.roomNumber;
    }

    public MapSite getSide(Direction direction) {
        return sides.get(direction);
    }

    public void setSide(Direction direction, MapSite side) {
        this.sides.put(direction, side);
    }


    @Override
    public void enter() {
        System.out.println("Entering regular room number: " + this.roomNumber);
    }

    @Override
    public String toString() {
        StringBuffer description = new StringBuffer();

        description.append("Regular room number: " + this.roomNumber + "\n");

        for (Direction dir : this.sides.keySet()) {
           description.append(" On the " + dir + " side, there is a " + this.sides.get(dir)+ "\n");
        }

        return description.toString();
    }
}

File: maze/Room.cpp
#include "Room.h"

string e_to_str(Direction dir) {
    switch(dir) {
        case East:
            return "East";
            break;
        case West:
            return "West";
            break;
        case North:
            return "North";
            break;
        case South:
            return "South";
            break;
        default:
            return "East";
    }
}

Room::Room(int roomNumber) {
    _roomNumber = roomNumber;
}

void Room::Enter() {
    cout << "Entered room number " << _roomNumber;
}

MapSite* Room::getSide(Direction direction) {
    return _sides[direction];
}

void Room::setSide(Direction direction, MapSite* side) {
    _sides[direction] = side;
}

void Room::printStatus() {
    cout << "for room " << _roomNumber << endl;
    Direction direction; 

    for(int i = 0; i < 4; i++) {
        direction = (Direction) i;
        cout << "a " << _sides[direction]->desc() << " in the " << e_to_str(direction) << endl;
    }
}

string Room::desc() {
    return "room";
}


